src/common/dao/project/projectmember.go

```diff
@@ -0,0 +1,121 @@
+
+import (
+	"fmt"
+
+	"github.com/vmware/harbor/src/common/dao"
+	"github.com/vmware/harbor/src/common/models"
+	"github.com/vmware/harbor/src/common/utils/log"
+)
+
+// GetProjectMember gets all members of the project.
+func GetProjectMember(queryMember models.Member) ([]*models.Member, error) {
+	log.Debugf("Query condition %+v", queryMember)
+	if queryMember.ProjectID == 0 {
+		return nil, fmt.Errorf("Failed to query project member, query condition %v", queryMember)
+	}
+
+	o := dao.GetOrmer()
+	sql := ` select a.* from ((select pm.id as id, pm.project_id as project_id, ug.id as entity_id, ug.group_name as entity_name, ug.creation_time, ug.update_time, r.name as rolename, 
+		r.role_id as role, pm.entity_type as entity_type from user_group ug join project_member pm 
+		on pm.project_id = ? and ug.id = pm.entity_id join role r on pm.role = r.role_id where  pm.entity_type = 'g')
+		union
+		(select pm.id as id, pm.project_id as project_id, u.user_id as entity_id, u.username as entity_name, u.creation_time, u.update_time, r.name as rolename, 
+		r.role_id as role, pm.entity_type as entity_type from user u join project_member pm 
+		on pm.project_id = ? and u.user_id = pm.entity_id 
+		join role r on pm.role = r.role_id where u.deleted = 0 and pm.entity_type = 'u')) as a where a.project_id = ? `
+
+	queryParam := make([]interface{}, 1)
+	// used ProjectID already
+	queryParam = append(queryParam, queryMember.ProjectID)
+	queryParam = append(queryParam, queryMember.ProjectID)
+	queryParam = append(queryParam, queryMember.ProjectID)
+
+	if len(queryMember.Entityname) > 0 {
+		sql += " and a.entity_name = ? "
+		queryParam = append(queryParam, queryMember.Entityname)
+	}
+
+	if len(queryMember.EntityType) == 1 {
+		sql += " and a.entity_type = ? "
+		queryParam = append(queryParam, queryMember.EntityType)
+	}
+
+	if queryMember.EntityID > 0 {
+		sql += " and a.entity_id = ? "
+		queryParam = append(queryParam, queryMember.EntityID)
+	}
+	if queryMember.ID > 0 {
+		sql += " and a.id = ? "
+		queryParam = append(queryParam, queryMember.ID)
+	}
+	sql += ` order by a.entity_name `
+	members := []*models.Member{}
+	_, err := o.Raw(sql, queryParam).QueryRows(&members)
+
+	return members, err
+}
+
+// AddProjectMember inserts a record to table project_member
+func AddProjectMember(member models.Member) (int, error) {
+
+	log.Debugf("Adding project member %+v", member)
+	o := dao.GetOrmer()
+
+	if member.EntityID <= 0 {
+		return 0, fmt.Errorf("Invalid entity_id, member: %+v", member)
+	}
+
+	if member.ProjectID <= 0 {
+		return 0, fmt.Errorf("Invalid project_id, member: %+v", member)
+	}
+
+	sql := "insert into project_member (project_id, entity_id , role, entity_type) values (?, ?, ?, ?)"
+	_, err := o.Raw(sql, member.ProjectID, member.EntityID, member.Role, member.EntityType).Exec()
+	if err != nil {
+		return 0, err
+	}
+	var pmid int
+	querySQL := "select id from project_member where project_id = ? and entity_id = ? and entity_type = ? limit 1"
```

Comment:
We can use sql.Result returned by Exec() to get the ID. Or use the beego ORM directly.

Reply:
Updated